crates/provider/src/layers/gas.rs

```diff
@@ -0,0 +1,296 @@
+    fn layer(&self, inner: P) -> Self::Provider {
+        GasEstimatorProvider { inner, _phantom: PhantomData }
+    }
+}
+
+/// A provider that estimates gas for transactions.
+///
+/// Note: This provider requires the chain_id to be set in the transaction request if it's a
+/// EIP1559.
+///
+/// You cannot construct this directly, use [`ProviderBuilder`] with a [`GasEstimatorLayer`].
+///
+/// [`ProviderBuilder`]: crate::ProviderBuilder
+#[derive(Debug, Clone)]
+pub struct GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    inner: P,
+    _phantom: PhantomData<(N, T)>,
+}
+
+impl<N, T, P> GasEstimatorProvider<N, T, P>
+where
+    N: Network,
+    T: Transport + Clone,
+    P: Provider<N, T>,
+{
+    /// Gets the gas_price to be used in legacy txs.
+    async fn get_gas_price(&self) -> TransportResult<U256> {
+        self.inner.get_gas_price().await
+    }
+
+    /// Gets the gas_limit to be used in txs.
+    async fn get_gas_estimate(&self, tx: &N::TransactionRequest) -> TransportResult<U256> {
+        self.inner.estimate_gas(tx, None).await
+    }
+
+    /// Gets the max_fee_per_gas and max_priority_fee_per_gas to be used in EIP-1559 txs.
+    async fn get_eip1559_fees_estimate(&self) -> TransportResult<Eip1559Estimation> {
+        self.inner.estimate_eip1559_fees(None).await
+    }
+
+    /// Populates the gas_limit, max_fee_per_gas and max_priority_fee_per_gas fields if unset.
+    /// Requires the chain_id to be set in the transaction request to be processed as a EIP-1559 tx.
+    /// If the network does not support EIP-1559, it will process it as a legacy tx.
+    async fn handle_eip1559_tx(
+        &self,
+        tx: &mut N::TransactionRequest,
+    ) -> Result<(), TransportError> {
+        let gas_estimate_fut = if let Some(gas_limit) = tx.gas_limit() {
+            async move { Ok(gas_limit) }.left_future()
+        } else {
+            async { self.get_gas_estimate(tx).await }.right_future()
+        };
+
+        let eip1559_fees_fut = if let (Some(max_fee_per_gas), Some(max_priority_fee_per_gas)) =
+            (tx.max_fee_per_gas(), tx.max_priority_fee_per_gas())
+        {
+            async move { Ok(Eip1559Estimation { max_fee_per_gas, max_priority_fee_per_gas }) }
+                .left_future()
+        } else {
+            async { self.get_eip1559_fees_estimate().await }.right_future()
+        };
+
+        match futures::try_join!(gas_estimate_fut, eip1559_fees_fut) {
+            Ok((gas_limit, eip1559_fees)) => {
+                tx.set_gas_limit(gas_limit);
+                tx.set_max_fee_per_gas(eip1559_fees.max_fee_per_gas);
+                tx.set_max_priority_fee_per_gas(eip1559_fees.max_priority_fee_per_gas);
+                Ok(())
+            }
+            Err(err) => {
+                if err.is_transport_error()
+                    && err.to_string() == *"EIP-1559 not activated".to_string()
```

Comment:
I'd love this to a helper method on the error re 1559 but we can do as followup

Reply:
actually this is a bit of a hole in the API. 1559 not activated should not be a transport error. "endpoint doesn't exist" falls in a hole between transport error and rpc error 🤔